src/main/java/org/opensearch/security/privileges/ResourceAccessEvaluator.java

```diff
@@ -0,0 +1,212 @@
+ *
+ * @opensearch.experimental
+ */
+public class ResourceAccessEvaluator {
+    private static final Logger log = LogManager.getLogger(ResourceAccessEvaluator.class);
+
+    private final Set<String> resourceIndices;
+    private final ThreadContext threadContext;
+    private final ResourceSharingIndexHandler resourceSharingIndexHandler;
+
+    public ResourceAccessEvaluator(
+        Set<String> resourceIndices,
+        ThreadPool threadPool,
+        ResourceSharingIndexHandler resourceSharingIndexHandler
+    ) {
+        this.resourceIndices = resourceIndices;
+        this.threadContext = threadPool.getThreadContext();
+        this.resourceSharingIndexHandler = resourceSharingIndexHandler;
+    }
+
+    /**
+     * Evaluate access to resources (example, docs in an index).
+     * The permissions will be evaluated based on the access-level the resource is shared at rather than roles that the requesting user is mapped to.
+     * This allows for a standalone authorization flow for users requesting access to resource.
+     * <p>
+     * 0. Creating a resource requires "create" permissions that are checked outside this evaluator.
+     * 1. Owners and admin-certificate users will be granted access automatically.
+     * 2. Even if a user has access to all indices, they will not be able to access a resource that they are not the owner of and is not shared with them.
+     * 3. A user with no index permissions may not be able to create a resource, however, they can modify and delete a resource shared with them at full-access level.
+     *
+     * @param request                         may contain information about the index and the resource being requested
+     * @param action                          the action being requested to be performed on the resource
+     * @param isResourceSharingFeatureEnabled flag to indicate whether this feature is enabled
+     * @param context                         the evaluation context to be used when performing authorization
+     * @param presponse                       the response which tells whether the action is allowed for user, or should the request be checked with another evaluator
+     * @return PrivilegesEvaluatorResponse may be complete if the request is for a resource and authz check was successful, incomplete otherwise
+     */
+    public PrivilegesEvaluatorResponse evaluate(
+        final ActionRequest request,
+        final String action,
+        boolean isResourceSharingFeatureEnabled,
+        final PrivilegesEvaluationContext context,
+        final PrivilegesEvaluatorResponse presponse
+    ) {
+
+        // If feature is disabled we skip evaluation through this evaluator
+        if (!isResourceSharingFeatureEnabled) {
+            return presponse;
+        }
+
+        // Skip evaluation if request is not a DocRequest type
+        if (!(request instanceof DocRequest req)) {
```

Comment:
Should we also add logic to skip DocWriteRequest?

Reply:
DocWriteRequest extends DocRequest class, so it will be skipped if I understand correctly